Fix error counting

nnethercote opened the pull request:
There is some messiness in how errors get counted. Here are some cleanups.

r? @compiler-errors

nnethercote committed:
Clarify comments about diagnostic count fields.

nnethercote committed:
Tweak error counting.

We have several methods indicating the presence of errors, lint errors,
and delayed bugs. I find it frustrating that it's very unclear which one
you should use in any particular spot. This commit attempts to instill a
basic principle of "use the least general one possible", because that
reflects reality in practice -- `has_errors` is the least general one
and has by far the most uses (esp. via `abort_if_errors`).

Specifics:
- Add some comments giving some usage guidelines.
- Prefer `has_errors` to comparing `err_count` to zero.
- Remove `has_errors_or_span_delayed_bugs` because it's a weird one: in
  the cases where we need to count delayed bugs, we should really be
  counting lint errors as well.
- Rename `is_compilation_going_to_fail` as
  `has_errors_or_lint_errors_or_span_delayed_bugs`, for consistency with
  `has_errors` and `has_errors_or_lint_errors`.
- Change a few other `has_errors_or_lint_errors` calls to `has_errors`,
  as per the "least general" principle.

This didn't turn out to be as neat as I hoped when I started, but I
think it's still an improvement.

nnethercote committed:
Count "unused extern" errors as lints rather than normal errors.

compiler-errors commented:
Looks good to me on a first pass, but given that this touches `track_errors` and other error-counting logic, maybe this could use a second review by oli

r? @oli-obk

oli-obk commented:
Most track_errors are getting removed in https://github.com/rust-lang/rust/pull/119869, the others in https://github.com/rust-lang/rust/pull/119895

Your solution looks simpler than mine, i'll have a detailed look at the differences tomorrow

bors commented:
:umbrella: The latest upstream changes (presumably #120112) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased and removed the "Remove `track_errors`" commit because #119869 landed in the meantime and it overlaps a great deal. (Which means the one use of `track_errors` is still present.)

@bors r=compiler-errors,oli-obk

bors commented:
:pushpin: Commit 1f9fa2305a099cbbf9858e3eb777495a7246e84a has been approved by `compiler-errors,oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors,oli-obk 1f9fa2305a099cbbf9858e3eb777495a7246e84a -->
<!-- homu: {"type":"Approved","sha":"1f9fa2305a099cbbf9858e3eb777495a7246e84a","approver":"compiler-errors,oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
